### ISSUE TITLE
Improve speed limit inference 

### ISSUE THREAD

**Issue by SiarheiFedartsou** [Project-OSRM/osrm-backend#6413]
# Description

There are a lot of cases when speed limits are not available as explicit `maxspeed` tag, but can be inferred from other tags(e.g. in Poland there is 50 km/h limit in urban area if there is no speed limit sign). We already partially handle such cases using tags [like `at:rural`](https://github.com/Project-OSRM/osrm-backend/blob/5e5f1f4add6d86b097569545b670de513ebc6b04/profiles/car.lua#L270), but it applies only if this tag explicitly present for way. At the same time there is this quite interesting tool https://github.com/westnordost/osm-legal-default-speeds which parses [OSM documentation](https://wiki.openstreetmap.org/wiki/Default_speed_limits) and then infers maxspeed by “implicit” tags. For example if we know that this particular road is in Belgium, there are no tags about max speed, but we have `highway=motorway` tag, we can infer 120 kmh limit, just because it is default limit for motorways in Belgium.

So the idea is that we could try to re-use this idea in OSRM in order to improve quality of max speed data we get from pbfs(it will be difficult to directly re-use tool I mentioned above though - it is written in Kotlin).

Video from the author of that tool https://www.youtube.com/watch?v=FH98z24PQGs&t=2s

I would appreciate any thoughts on this.

**Comment by mjjbell** [comment]
Could extend the [geojson data](https://github.com/Project-OSRM/osrm-backend/tree/master/data) functionality to set max speeds in a similar way.

**Comment by jcoupey** [comment]
Worth noting there is also [this approach](https://github.com/OpenStreetMapSpeeds/conflation) followed by the Valhalla team as an independent project. My understanding is that the scope is a bit different as the aim is to derive typical expected speeds from GPS traces rather than "just" legal max speed values.

**Comment by nilsnolde** [comment]
There's also this discussion over in the Valhalla repo in case it's of interest: https://github.com/valhalla/valhalla/issues/3334#issuecomment-1261602765. 

@SiarheiFedartsou 
> it will be difficult to directly re-use tool I mentioned above though - it is written in Kotlin

It's apparently a "multiplatform" library [according to the author](https://github.com/valhalla/valhalla/issues/3334#issuecomment-1203257035), and as I understand you can use that directly in C++ (via `extern C` blocks). I have never done that myself either, guess/hope one finds easy examples online.

> as the aim is to derive typical expected speeds from GPS traces rather than "just" legal max speed values

I think both properties/projects are nice to have. `OpenStreetMapSpeeds` describes a more realistic speed that you'd probably rather use for costing/ETA (with density indication in the JSON schema Valhalla can make use of, I guess OSRM not yet). @westnordost 's returns very detailed legal speed limit tags, which also have important use cases for navigation signage/maneuver instructions or fill-in lacking more realistic speeds, also possibly for costing/restrictions (though no one uses it that way yet I think..). It'd be super nice if one could use one of the projects at runtime, so the graph wouldn't have to store so many speed values. Maybe e.g. storing the `OpenStreetMapSpeeds` more realistic ones in the graph and having `osm-legal-default-speeds` fill in the holes with legal speed limits and/or use it to set a legal speed limit for maneuver instructions. Maybe that's asking for too much though, not sure what the performance penalty would be to have the Kotlin lookup during the algo's expansion (though it seems to be pretty simple in-memory lookup). What do you think?

FWIW, Valhalla already uses `OpenStreetMapSpeeds`, but the coverage is very flaky so far. I'd like to find some time to try out an implementation of `osm-legal-default-speeds` (and/or improve OpenStreetMapSpeeds), but can't see that happening in the next weeks. Anyways, I'll watch here and update if I/we should try it before you do.

Also, I just now had the time to look in detail into `osm-legal-default-speeds`, and wow, really tons of work! Nice wiki too. Congrats @westnordost !

**Comment by SiarheiFedartsou** [comment]
> Regarding Kotlin Multiplatform - I haven't used it either except for running the unit tests in native. I've also not seen (or searched for) any examples how to include kotlin multiplatform dependencies into a C/C++ project.

It definitely should be possible - I know people who did something like this, but I guess we will have to add Kotlin runtime as a dependency - not sure we want to do that 🤔 

@westnordost as author of this library what do you think will it be difficult to port it to C++? I briefly looked over the code and it seems to be relatively simple, but may be I didn't notice something?

**Comment by westnordost** [comment]
It's about 1000 SLOC, not counting the unit tests. (Kotlin code is quite concise however, probably you should count with 1500 SLOC in C++)

**Comment by westnordost** [comment]
But ofc I would advise against it as long as the Kotlin Native stuff is not completely unusable. It then becomes double the effort to maintain. I found this in the Kotlin docs: https://kotlinlang.org/docs/native-dynamic-libraries.html#compile-and-run-the-example-on-linux-and-macos

**Comment by nilsnolde** [comment]
It's getting a bit off-track, but just thinking about how to integrate that project while building/linking.. To reduce duplication across repos, either a CMakeLists.txt for the kotlin project and/or a (external?) conan recipe could be good right? Both OSRM & Valhalla are using conan already, so minimal effort once that'd stand. Never wrote a recipe before and no idea what's lurking when writing a stable recipe for most archs/platforms. Also the example uses gradle, so that's smth a user who wants to build this locally, would need.. That does sound a little detracting, at least to me.. Only alternative I can think of is to expect the user to install the project system-wide and let CMake find it, not super nice either, sounds like potentially many annoyed (& annoying) bug reports.

Any other thoughts on integration? Maybe I'm thinking too complex..

**Comment by westnordost** [comment]
Sorry, I have no knowledge of the project setup / build pipeline stuff in the C/C++ world. If the linked docs above do not help (then they should be improved and) then I guess it can only help to find some examples of kotlin multiplatform libs used in native projects on the internet.

**Comment by nilsnolde** [comment]
Maybe @SiarheiFedartsou has a point and it might be easier for us to use when it's ported into a e.g. header-only C++ lib that we all could use. I don't think this has to be a linked library, just complicates it quite a bit, apart from all the basics like having gradle installed. Header-only would be super easy to integrate. Maybe we don't even need some of the more advanced functions with fuzzy matching, not sure though. The serious work is probably more in the JSON you put together, writing a tiny C++ parser for it might be less work than extending the build pipeline for gradle/kotlin.

**Comment by westnordost** [comment]
The major part of the library actually consists of parsing the tag filter syntax as is used here: 
https://wiki.openstreetmap.org/wiki/Default_speed_limits#Road_types_to_tag_filters

**Comment by SiarheiFedartsou** [comment]
@westnordost yep, but we still can use this part of the library, but only have "matching" part rewritten in C++.

**Comment by westnordost** [comment]
Not sure what you mean. That code,  "as is", is written in kotlin. Why would you be able to use that code written in kotlin but not the other part written in kotlin?

**Comment by westnordost** [comment]
> tbh not sure about tag filter expressions if it will be needed or not.

It is needed. Otherwise no matching of OSM tags to road types is possible.

**Comment by DennisOSRM** [comment]
I think having something like this would be a nice addition.

**Comment by ftrebien** [comment]
An idea: once thinking about the Valhalla algorithm, I obtained blob for urban areas in GIMP using simple median filters. I started with the streets rendered as thick black lines (eg. [Stamen's Toner](http://maps.stamen.com/toner/)) against a white background, filtered with settings to expand them until they merge into on big blob per city, then filtered again with roughly inverse settings to contract some more than the same amount so that the roads between cities to disappear, leaving city blobs. Custom rendering rules would allow skipping the first filter if the streets were rendered thick enough for them to merge. The missing (though apparently easy) step would be vectorization, which would produce a geometry that could be used to check, during map processing, whether any roads are inside or outside a built-up area such as a city. I haven't investigated, but it may be possible to do all this with just vectors (instead of raster images) in QGIS using the buffer tool. City boundaries don't change much, so built area boundaries may be updated infrequently, say, once a year.
